Add more comprehensive icmp6 snat testing

tommyp1ckles opened the pull request:
ICMPv6 PMTU response datapath logic is somewhat unique in that it depends in parsing encapsulated l4 headers
after the icmp header.   This is prone to regressions as the functionality may not be regularly exercised by normal e2e tests.

This refactors the bpf_nat6_netdev_tests bpf tests into a header `bpf_nat_icmp6.h` which instead provides a macro that expands into tcp/udp tests. As well, as adding a sctp scenario to provide coverage for all available l4 protocols.

Using this, we create three new test files to prevent regression on several env configuration tuples:

* tunnel=false hostfw=false
* tunnel=false hostfw=true
* tunnel=true hostfw=false

I'm also currently looking at possibly expanding E2E testing,  but this is a bit tricky due to needing some kind of external network upon which to test the "bounced" mtu messages.  Therefore, this should provide some protection against regressions in the meantime.

Also, we may expand this as needed to reach the desired coverage by adding further configuration tuples (**note:** we may also want to add some test cases covering non-"pkt-too-big" icmp6 cases).

ldelossa commented on bpf/tests/bpf_nat_icmp6.h:
```diff
@@ -0,0 +1,366 @@
+ */
+__always_inline int gen_pmtu_pkt(struct pktgen *builder, int l4_type)
+{
+	struct ethhdr *l2;
```
Maybe you want to define all these as NIL. Per the C spec, these stack vars may contain garbage.

ldelossa commented on bpf/tests/bpf_nat_icmp6.h:
```diff
@@ -0,0 +1,366 @@
+
+#define PMTU_TEST(name)				\
+						\
+struct {					\
```
This amount of logic written in macros is a bit brutal. 

I see you do this just to expand the macros in each test case. 

Would it be possible to get the same amount of abstraction you want here by using minimal macros and shared C functions in this header? 

I don't necessarily see much conditional macro expansion based on the 'name' of the test.

tommyp1ckles replied:
yeah good call, ill rework this 👍

tommyp1ckles replied:
@ldelossa  Reworked this, moving most of the reused code into functions.  I left one macro in there called CHECK_PMTU_SNAT, just to inject the identical test code into various test cases while also having access to some of the test cursor variables that are brought into scope via `test_init()`.

I could put this into a function, but I'd have to flatten the errors returned not being able to call test_fatal which would make these more difficult to debug. 

let me know what you think

tommyp1ckles committed:
bpf/tests: expand coverage for bpf icmp6 nodeport snat tests.

ICMPv6 PMTU response datapath logic is somewhat unique in that it
depends in parsing encapsulated l4 headers after the icmp header.
This PR expands the coverage for logic covering the scenario of

This is prone to regressions as the functionality may not be
regularly exercised by normal e2e tests.

This refactors the bpf_nat6_netdev_tests bpf tests into a header
`bpf_nat_icmp6.h` which instead provides a macro that expands into
tcp/udp tests. As well, as adding a sctp scenario to provide
coverage for all available l4 protocols.

Using this, we create three new test files to prevent regression
on several env configuration tuples:

* tunnel=false hostfw=false
* tunnel=false hostfw=true
* tunnel=true hostfw=false

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

ldelossa commented on bpf/tests/bpf_nat_icmp6.h:
```diff
@@ -256,49 +182,38 @@ int snat_v6_pmtu_udp_setup(struct __ctx_buff *ctx)
+	return map_update_elem(&cilium_snat_v6_external, &tuple, &entry, BPF_ANY);
 }
 
+#define CHECK_PMTU_SNAT()									\
```
I personally would take the body of this, and make it a function (or inline function), and then pass arguments into CHECK_PMTU_SNAT and call the function with those arguments, so it is clear what values the macro is operating on, and also gets rid of more macro content which tools and LSPs don't parse correctly.

tommyp1ckles replied:
Yeah I think I've come around to the same view, I wanted to macro to be able to make use of the test fixtures as a inline function won't have those in scope like the macro code would.

My thinking was, this would just make it a bit more clear where errors where happening, but I think the extra macro code will ultimately override any benefit there.

I'll make the changes and push it up

ldelossa reviewed (approved):
I have a further comment about the CHECK_PMTU_SNAT macro, but this is a big improvement. Going to approve. You can consider my suggestion if you agree.
